### DISCUSSION TITLE
Error compiling LIS

### DISCUSSION THREAD

**Discussion by laurizioA** [NASA-LIS/LISF#1189] [Q&A] [answered]
Hi!

An error occurs when compiling the LIS, the following message appears:

**"...
Generating dependencies for ../dataassim/obs/ANSA_SNWD/ANSASNWDsnow_Mod.F90
Generating dependencies for ../dataassim/obs/ANSA_SCF/ANSASCFsnow_Mod.F90
Generating dependencies for ../optUE/type/paramestim/obs/AMSRE_SR/AMSRE_SR_em_obsMod.F90
gfortran -I/home/codesal/anaconda3/envs/lisf/include/ -I/home/codesal/anaconda3/envs/lis/include/ -I/home/codesal/anaconda3/envs/lis/include/ -I/home/codesal/anaconda3/envs/lis/include/ -I/home/codesal/anaconda3/envs/lis/include/ -DLIS_JULES -I./ -I../core/ -I../params/gfrac/SPORTDaily/ -I../params/gfrac/VIIRSDaily/ -I../surfacemodels/land/vic.4.1.1/ -I../surfacemodels/land/vic.4.1.1/physics/ -I../surfacemodels/land/vic.4.1.2.l/ -I../surfacemodels/land/vic.4.1.2.l/physics/ -I../surfacemodels/land/awral.6.0.0/physics/ -I../surfacemodels/land/rdhm.3.5.6/ -I../surfacemodels/land/rdhm.3.5.6/sac/ -I../surfacemodels/land/rdhm.3.5.6/frz/ -I../surfacemodels/land/rdhm.3.5.6/snow17/ -I../surfacemodels/land/rdhm.3.5.6/sachtet/ ../core/LIS_mpiMod.F90
/usr/lib/gcc/x86_64-linux-gnu/7/../../../x86_64-linux-gnu/Scrt1.o: In function "_start":
(.text+0x20): undefined reference to "main"
collect2: error: ld returned 1 exit status
Makefile:140: recipe for target 'LIS_mpiMod.o' failed
make: *** [LIS_mpiMod.o] Error 1

[ERR] Compile failed
-------------------------------------------------------------------"**

Could anyone help in solving this problem?

**Comment by dmocko** [comment, marked as answer]
Hi @laurizioA 

Sorry for the delay in the response.  Several team members are getting the last of summer vacations in before schools start up.

Can you please attach your "configure.lis" file to this thread?

**Reply by laurizioA** [reply]
Thank you @dmocko 

Follow the configure.lis file:
[configure.lis.txt](https://github.com/NASA-LIS/LISF/files/9427195/configure.lis.txt)

**Reply by dmocko** [reply]
Got it, thanks.

What operating system are you using?  Also, did you compile ESMF with MPI?

If you can, please also attach the result of your $PATH and your $LD_LIBRARY_PATH.

We will look into this, but nothing has popped right away.

**Reply by laurizioA** [reply]
What operating system are you using? 
**Ubuntu 20.04.4 LTS**

Also, did you compile ESMF with MPI? 
**I installed ESMF in the Anaconda3 virtual environment. Together, the mpi (v1.0) and mpich (v4.0.2) package was installed.**

If you can, please also attach the result of your $PATH and your $LD_LIBRARY_PATH.
**I'll do that, I haven't done it before.**

I will try now.

Thanks!

**Reply by laurizioA** [reply]
The previous compilation issue has been resolved but another error has occurred.

**gfortran -c -ffree-line-length-0 -O2  -fbacktrace -DSPMD -fconvert=big-endian -DHIDE_SHR_MSG -DNO_SHR_VMATH -DGFORTRAN -DLINUX  -I/home/codesal/anaconda3/envs/lisf/mod -I/home/codesal/anaconda3/envs/lisf//include/ -I/home/codesal/anaconda3/envs/lisf//include/ -I/home/codesal/anaconda3/envs/lisf//include/ -I/home/codesal/anaconda3/envs/lisf//include/ -I/home/codesal/anaconda3/envs/lisf//include/ -DLIS_JULES -I./ -I../core/ -I../params/gfrac/SPORTDaily/ -I../params/gfrac/VIIRSDaily/ -I../surfacemodels/land/vic.4.1.1/ -I../surfacemodels/land/vic.4.1.1/physics/ -I../surfacemodels/land/vic.4.1.2.l/ -I../surfacemodels/land/vic.4.1.2.l/physics/ -I../surfacemodels/land/awral.6.0.0/physics/ -I../surfacemodels/land/rdhm.3.5.6/ -I../surfacemodels/land/rdhm.3.5.6/sac/ -I../surfacemodels/land/rdhm.3.5.6/frz/ -I../surfacemodels/land/rdhm.3.5.6/snow17/ -I../surfacemodels/land/rdhm.3.5.6/sachtet/ ../dataassim/obs/ANSA_SNWD/read_ANSASNWDsnow.F90
../dataassim/obs/ANSA_SNWD/read_ANSASNWDsnow.F90:611:2:

  611 | #endif
      |  1~~~~    
Fatal Error: hdf.f90: Non-existent file or directory
compilation terminated.
make: *** [Makefile:140: read_ANSASNWDsnow.o] Erro 1

[ERR] Compile failed
-------------------------------------------------------------------**

I didn't find the file hdf.f90 in the path of "configure.lis" of LIS_HDF4. 
HDF4 was installed via Anaconda3.

**Reply by dmocko** [reply]
The hdf.f90 file should here for you:  /home/codesal/anaconda3/envs/lisf/include/hdf.f90

If it's not there, then you will need to re-install HDF4.

Please be sure that this is in your LD_LIBRARY_PATH:  /home/codesal/anaconda3/envs/lisf/lib/

**Reply by laurizioA** [reply]
Resolved previous issues, but other errors appear.
Included in .bashrc:
**export PATH=/home/codesal/anaconda3/envs/lisf:$PATH
export LD_LIBRARY_PATH=/usr/local/hdf4/lib:/home/codesal/anaconda3/envs/lisf/lib:$LD_LIBRARY_PATH**

When I run the .compile the following error appears:

**gfortran -c -ffree-line-length-0 -O2  -fbacktrace -DSPMD -fconvert=big-endian -DHIDE_SHR_MSG -DNO_SHR_VMATH -DGFORTRAN -DLINUX  -I/home/codesal/anaconda3/envs/lisf/mod -I/home/codesal/anaconda3/envs/lisf/include/ -I/home/codesal/anaconda3/envs/lisf/include/ -I/home/codesal/anaconda3/envs/lisf/include/ -I/usr/local/hdf4/include/ -I/home/codesal/anaconda3/envs/lisf/include/ -DLIS_JULES -I./ -I../core/ -I../params/gfrac/SPORTDaily/ -I../params/gfrac/VIIRSDaily/ -I../surfacemodels/land/vic.4.1.1/ -I../surfacemodels/land/vic.4.1.1/physics/ -I../surfacemodels/land/vic.4.1.2.l/ -I../surfacemodels/land/vic.4.1.2.l/physics/ -I../surfacemodels/land/awral.6.0.0/physics/ -I../surfacemodels/land/rdhm.3.5.6/ -I../surfacemodels/land/rdhm.3.5.6/sac/ -I../surfacemodels/land/rdhm.3.5.6/frz/ -I../surfacemodels/land/rdhm.3.5.6/snow17/ -I../surfacemodels/land/rdhm.3.5.6/sachtet/ ../core/LIS_DAobservationsMod.F90
../core/LIS_DAobservationsMod.F90:1327:18:

 1327 |    call MPI_BCAST(LIS_obs_offsets(k,:), LIS_npes, MPI_INTEGER,0, LIS_mpi_comm, ierr)
      |                  1
......
 1822 |    call MPI_BCAST(gtmp1d, LIS_rc%obs_gnc(k)*LIS_rc%obs_gnr(k), MPI_REAL,0, &
      |                  2
Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(4)/REAL(4)).
../core/LIS_DAobservationsMod.F90:1168:25:

 1168 |           call MPI_BCAST(LIS_obs_goffsets(k,:), LIS_npes, MPI_INTEGER,0, &
      |                         1
......
 1822 |    call MPI_BCAST(gtmp1d, LIS_rc%obs_gnc(k)*LIS_rc%obs_gnr(k), MPI_REAL,0, &
      |                  2       
Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(4)/REAL(4)).
../core/LIS_DAobservationsMod.F90:1192:25:

 1192 |           call MPI_BCAST(LIS_obsens_goffsets(k,:), LIS_npes, MPI_INTEGER,0, &
      |                         1
......
 1822 |    call MPI_BCAST(gtmp1d, LIS_rc%obs_gnc(k)*LIS_rc%obs_gnr(k), MPI_REAL,0, &
      |                  2       
Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(4)/REAL(4)).
../core/LIS_DAobservationsMod.F90:369:25:

  369 |           call MPI_BCAST(LIS_ooffsets(k,:), LIS_npes, MPI_INTEGER,0, &
      |                         1
......
 1822 |    call MPI_BCAST(gtmp1d, LIS_rc%obs_gnc(k)*LIS_rc%obs_gnr(k), MPI_REAL,0, &
      |                  2       
Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(4)/REAL(4)).
make: *** [Makefile:140: LIS_DAobservationsMod.o] Erro 1

[ERR] Compile failed
-------------------------------------------------------------------**

The installed MPI is version 4.0.2.

**Reply by Lise-Li** [reply]
Hi, @laurizioA have your problem solved? I met the same problem with you, and I don't know how to solve it. Would you mind letting me know how you configured HDF4? Thank you!

**Reply by laurizioA** [reply]
Hi, @Lise-Li!

I installed hdf4 by "sudo apt install hdf4-tools -y" and pointed to "export LDT_HDF4=/usr". But I have a problem with ESMF.

**Reply by Lise-Li** [reply]
Hi,@laurizioA! Thank you for your reply! I'll try again!
